plugins/receivers/lark/client.go

```diff
@@ -0,0 +1,223 @@
+	larkPathOAuth      = "/open-apis/auth/v3/tenant_access_token/internal/"
+)
+
+type Channel struct {
+	ID   string `json:"id"`
+	Name string `json:"name"`
+}
+
+type ClientOption func(*Client)
+
+// ClientWithHTTPClient assigns custom http client when creating a lark client
+func ClientWithHTTPClient(httpClient *httpclient.Client) ClientOption {
+	return func(c *Client) {
+		c.httpClient = httpClient
+	}
+}
+
+// ClientWithRetrier wraps client call with retrier
+func ClientWithRetrier(runner retry.Runner) ClientOption {
+	return func(c *Client) {
+		// note: for now retry only happen in send message context method
+		c.retrier = runner
+	}
+}
+
+type Client struct {
+	cfg        AppConfig
+	httpClient *httpclient.Client
+	retrier    retry.Runner
+}
+
+// NewClient is a constructor to create lark client.
+// this version uses lark v3 SDK.
+func NewClient(cfg AppConfig, opts ...ClientOption) *Client {
+	c := &Client{
+		cfg: cfg,
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	if cfg.APIHost == "" {
+		c.cfg.APIHost = defaultLarkAPIHost
+	}
+
+	// sanitize
+	c.cfg.APIHost = c.cfg.APIHost + "/"
+
+	if c.httpClient == nil {
+		c.httpClient = httpclient.New(cfg.HTTPClient)
+	}
+
+	return c
+}
+
+// GetWorkspaceChannels fetches list of joined channel of a client
+func (c *Client) GetWorkspaceChannels(ctx context.Context, clientID, clientSecret string) ([]Channel, error) {
+	var client = lark.NewClient(clientID, clientSecret)
+
+	joinedChannelList, err := c.getJoinedChannelsList(ctx, client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch joined channel list: %w", err)
+	}
+
+	result := make([]Channel, 0)
+	for _, c := range joinedChannelList {
+		result = append(result, Channel{
+			ID:   *c.ChatId,
+			Name: *c.Name,
+		})
+	}
+	return result, nil
+}
+
+func (c *Client) Notify(ctx context.Context, conf NotificationConfig, message Message) error {
+	if c.retrier != nil {
+		if err := c.retrier.Run(ctx, func(ctx context.Context) error {
+			return c.notify(ctx, conf, message)
+		}); err != nil {
+			return err
+		}
+	}
+	return c.notify(ctx, conf, message)
+}
+
+// Notify sends message to a specific lark channel
+func (c *Client) notify(ctx context.Context, conf NotificationConfig, message Message) error {
+
+	var client = lark.NewClient(conf.ClientID, conf.ClientSecret)
+
+	var channelID string
+	switch conf.ChannelType {
+	case TypeChannelChannel:
+		joinedChannelList, err := c.getJoinedChannelsList(ctx, client)
+		if err != nil {
+			if err := c.checkLarkErrorRetryable(err); errors.As(err, new(retry.RetryableError)) {
+				return err
+			}
+			return fmt.Errorf("failed to fetch joined channel list: %w", err)
+		}
+		channelID = searchChannelId(joinedChannelList, message.Channel)
+		if channelID == "" {
+			return fmt.Errorf("app is not part of the channel %q", message.Channel)
+		}
+	case TypeChannelUser:
+		user, err := c.getUserByEmail(ctx, message.Channel, client)
+		if err != nil {
+			if err.Error() == "users_not_found" {
```

Comment:
Is Lark SDK/API return the same error string like this in case user not found?

Reply:
returning the error from the method getUserByEmail. lark API gives 200 always